### ISSUE TITLE
grafana_role_assignment unexpected behaviour - only possible to use a role only once?

### ISSUE THREAD

**Issue by Gammenon** [grafana/terraform-provider-grafana#1000]
Hi,

first of all I'm not sure if this is a Bug or just an unexpected/inconsistent behaviour in regards to the API.
So pardon me if this just works as designed.

### Terraform Version
* Terraform: v1.5.4
* Terraform Grafana Provider: 1.43.0
* Grafana: 10.1.0

### Affected Resource(s)
Please list the resources as a list, for example:
* grafana_role_assignment

### Expected Behavior
I feed the creation of grafana teams from a json file. The json object per team contains the team information itself as well as certain rbac role that should be applied.
I would have expected that it behaves similiar to the api command where I give a team and tell which roles should be applied

### Actual Behavior
If I understand right what happens here, grafana_role_assignment wants a role uid and a list of e.g. teams that should have this role.
through the loop over the json objects the same role_uid is used multiple times but it seems that the only one call of a role assignment per role is possible. every other will overwrite stuff. The result is, each plan/apply will switch forth an back the team id attached to the role
```
###############################################################
# First run

# module.grafana_teams["TeamA"].grafana_role_assignment.this["alert full access"] will be updated in-place
  ~ resource "grafana_role_assignment" "this" {
        id               = "Ev5RTDQ7z"
      ~ teams            = [
          - 336,
          + 344,
        ]
        # (3 unchanged attributes hidden)
    }

  # module.grafana_teams["TeamA"].grafana_role_assignment.this["report writer"] will be updated in-place
  ~ resource "grafana_role_assignment" "this" {
        id               = "D6I5sVE7k"
      ~ teams            = [
          - 339,
          + 344,
        ]
        # (3 unchanged attributes hidden)
    }


###############################################################
# Second run

  # module.grafana_teams["TeamB"].grafana_role_assignment.this["alert full access"] will be updated in-place
  ~ resource "grafana_role_assignment" "this" {
        id               = "Ev5RTDQ7z"
      ~ teams            = [
          - 344,
          + 336,
        ]
        # (3 unchanged attributes hidden)
    }

  # module.grafana_teams["TeamB"].grafana_role_assignment.this["report reader"] will be updated in-place
  ~ resource "grafana_role_assignment" "this" {
        id               = "D6I5sVE7k"
      ~ teams            = [
          - 344,
          + 339,
        ]
        # (3 unchanged attributes hidden)
    }
```

For me and at least for my provisioning this is pretty uncomfortable and seems to behave different than the API and as with similiar topics with other resources (e.g. datasource permissions)

**Comment by julienduchesne** [comment]
That is how the resource works for now. It takes over the role's assignments completely. 

Something could definitely be done to go with what other providers do: 
1. A block-type attribute within the role resource for cases where it's centrally managed
2. a separate resource that manages a single role assignment for cases where the role is created somewhere and assignees elsewhere

That would make more sense than the current 1:1 resource

**Comment by Gammenon** [comment]
Hi,

thanks for the answer.
from a provisioning point of view the current implementation does not make sense.
From my perspective I will deploy certain teams and want to apply certain rules to them.
Currently I would need to maintain my team related data in different places or have a much more complicated setup for the deployment.

In addition to this it is not aligned with the API behaviour which is a little bit confusing and makes it harder to switch from API provisioning to a (in general more comfortable) terraform deployment.

Could you give any idea of how long it would take to implement this and what would be the way to have this feature implemented at all?

**Comment by Gammenon** [comment]
Hi @julienduchesne,

I'm as well already in contact with the support, but wanted to add this here as well for the record if somebody comes across the same issue.

I just reworked my provisioning around this discussed behaviour.
The conclusion is devastating from my point of view.
Just in the moment after implementation where I thought, OK we handle it with this behaviour, I recognized that if I do a role assignment on that way it is absolute.
Every other role assignment, manual or via API, is thereby cancelled and overwritten which render this terraform method useless when you need to be able to use manual and provisioning mode.

**Comment by ImTomEddy** [comment]
I've just stumbled upon this ticket after spending an embarrassing amount of time debugging an issue which is caused by the ability to only provision one role assignment per role.

Would it be possible to get the documentation updated to clarify this as a temporary measure while the new feature is still in planning/progress?

**Comment by julienduchesne** [comment]
I am actually planning on working on this issue this week. I hope to be able to add new resources to add single rules (rather manage the whole ruleset)

**Comment by ImTomEddy** [comment]
@julienduchesne I've just started to implement the `grafana_role_assignment_item` resource you've added but upon investigation it has similar behaviour to the previous `grafana_role_assignment` resource and it will override any manually provisioned assignments for the given role due to the endpoint being used (`PUT /access-control/roles/{roleUID}/assignments`).

Is there a plan to implement individual role assignments based on the following endpoints:
- https://grafana.com/docs/grafana/latest/developers/http_api/access_control/#create-and-remove-team-role-assignments
- https://grafana.com/docs/grafana/latest/developers/http_api/access_control/#create-and-remove-service-account-role-assignments
- https://grafana.com/docs/grafana/latest/developers/http_api/access_control/#create-and-remove-user-role-assignments

As these would allow for individual assignments to be managed without the possibility of mutating external state.

If you think it's worth raising a separate ticket, please give me a shout 😄
